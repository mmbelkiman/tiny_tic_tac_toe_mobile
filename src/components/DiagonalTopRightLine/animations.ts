import {Animated} from 'react-native';

export const playHeightAnimation = (
  animation: Animated.Value,
  height: number,
) => {
  Animated.timing(animation, {
    toValue: height,
    duration: 80,
    useNativeDriver: false,
  }).start();
};

export const playShakeAnimation = (animation: Animated.Value) => {
  Animated.sequence([
    Animated.timing(animation, {
      toValue: 1,
      duration: 200,
      useNativeDriver: false,
    }),
    Animated.timing(animation, {
      toValue: 0,
      duration: 200,
      useNativeDriver: false,
    }),
  ]).start();
};

export const getShakeStyle = (
  animation: Animated.Value,
  length: number,
  height: number,
) => ({
  transform: [
    {translateX: length / 2},
    {translateY: -height / 2},
    {rotate: '-45deg'},
    {translateX: -length / 2},
    {translateY: height / 2},
    {
      translateY: animation.interpolate({
        inputRange: [0, 1],
        outputRange: [0, -5],
      }),
    },
  ],
});
