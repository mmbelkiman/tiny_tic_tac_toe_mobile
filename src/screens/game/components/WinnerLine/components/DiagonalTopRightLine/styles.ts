import {StyleSheet} from 'react-native';

export default StyleSheet.create({
  line: {
    position: 'absolute',
    borderRadius: 10,
  },
});
