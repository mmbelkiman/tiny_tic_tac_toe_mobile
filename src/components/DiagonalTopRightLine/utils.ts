import {getPlayerColor} from '../../common/utils';
import styles from './styles';
import {getShakeStyle} from './animations';
import {Animated} from 'react-native';
import {PlayerMark} from '../../common/types';

export const getLineStyle = ({
  winner,
  length,
  height,
  right,
  heightAnimation,
  opacityAnimation,
  shakeAnimation,
}: {
  winner: PlayerMark;
  length: number;
  height: number;
  right: number;
  heightAnimation: Animated.Value;
  opacityAnimation: Animated.Value;
  shakeAnimation: Animated.Value;
}) => [
  styles.line,
  {
    backgroundColor: getPlayerColor(winner),
    width: length,
    right: right,
    height: heightAnimation,
    opacity: opacityAnimation,
  },
  getShakeStyle(shakeAnimation, length, height),
];
