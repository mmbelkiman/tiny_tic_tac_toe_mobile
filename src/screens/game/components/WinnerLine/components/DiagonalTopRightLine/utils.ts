import {Animated} from 'react-native';
import styles from './styles';

export const getShakeStyle = (
  animation: Animated.Value,
  length: number,
  height: number,
) => ({
  transform: [
    {translateX: length / 2},
    {translateY: -height / 2},
    {rotate: '-45deg'},
    {translateX: -length / 2},
    {translateY: height / 2},
    {
      translateY: animation.interpolate({
        inputRange: [0, 1],
        outputRange: [0, -5],
      }),
    },
  ],
});

export const getLineStyle = ({
  color,
  length,
  height,
  right,
  heightAnimation,
  opacityAnimation,
  shakeAnimation,
}: {
  color: string;
  length: number;
  height: number;
  right: number;
  heightAnimation: Animated.Value;
  opacityAnimation: Animated.Value;
  shakeAnimation: Animated.Value;
}) => [
  styles.line,
  {
    backgroundColor: color,
    width: length,
    right: right,
    height: heightAnimation,
    opacity: opacityAnimation,
  },
  getShakeStyle(shakeAnimation, length, height),
];
