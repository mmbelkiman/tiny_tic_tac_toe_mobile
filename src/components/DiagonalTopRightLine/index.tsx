import React, {useEffect, useRef} from 'react';
import {Animated} from 'react-native';
import {PlayerMark} from '@/common/types';
import {playOpacityAnimation} from '@/common/animations';
import {getLineStyle} from './utils';
import {playHeightAnimation, playShakeAnimation} from './animations';

interface DiagonalTopRightLineProps {
  length: number;
  height: number;
  winner: PlayerMark | null;
  visible: boolean;
}

const DiagonalTopRightLine: React.FC<DiagonalTopRightLineProps> = ({
  length,
  height,
  winner,
  visible,
}) => {
  const opacityAnimation = useRef(new Animated.Value(0)).current;
  const heightAnimation = useRef(new Animated.Value(0)).current;
  const shakeAnimation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    playOpacityAnimation(opacityAnimation);
    playHeightAnimation(heightAnimation, height);
    playShakeAnimation(shakeAnimation);
  }, [opacityAnimation, heightAnimation, shakeAnimation, height]);

  if (!visible || winner === null) {
    return null;
  }

  return (
    <Animated.View
      style={getLineStyle({
        winner,
        length,
        height,
        right: height,
        heightAnimation,
        opacityAnimation,
        shakeAnimation,
      })}
    />
  );
};

export default DiagonalTopRightLine;
