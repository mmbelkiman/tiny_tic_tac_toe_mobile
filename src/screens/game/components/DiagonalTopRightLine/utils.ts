import {Animated} from 'react-native';
import {PlayerMark} from '../../../../common/types';
import {getPlayerColor} from '../../common/utils';
import styles from './styles';

export const getShakeStyle = (
  animation: Animated.Value,
  length: number,
  height: number,
) => ({
  transform: [
    {translateX: length / 2},
    {translateY: -height / 2},
    {rotate: '-45deg'},
    {translateX: -length / 2},
    {translateY: height / 2},
    {
      translateY: animation.interpolate({
        inputRange: [0, 1],
        outputRange: [0, -5],
      }),
    },
  ],
});

export const getLineStyle = ({
  winner,
  length,
  height,
  right,
  heightAnimation,
  opacityAnimation,
  shakeAnimation,
}: {
  winner: PlayerMark;
  length: number;
  height: number;
  right: number;
  heightAnimation: Animated.Value;
  opacityAnimation: Animated.Value;
  shakeAnimation: Animated.Value;
}) => [
  styles.line,
  {
    backgroundColor: getPlayerColor(winner),
    width: length,
    right: right,
    height: heightAnimation,
    opacity: opacityAnimation,
  },
  getShakeStyle(shakeAnimation, length, height),
];
