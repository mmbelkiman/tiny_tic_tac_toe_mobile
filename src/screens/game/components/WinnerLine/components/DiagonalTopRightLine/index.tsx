import React, {useEffect, useRef} from 'react';
import {Animated} from 'react-native';
import {
  playOpacityAnimation,
  playShakeAnimation,
  playSizeAnimation,
} from '../../../../common/animations';
import {getLineStyle} from './utils';
import {PlayerMark} from '../../../../common/types';

interface DiagonalTopRightLineProps {
  length: number;
  height: number;
  winner: PlayerMark;
}

const DiagonalTopRightLine: React.FC<DiagonalTopRightLineProps> = ({
  length,
  height,
  winner,
}) => {
  const opacityAnimation = useRef(new Animated.Value(0)).current;
  const heightAnimation = useRef(new Animated.Value(0)).current;
  const shakeAnimation = useRef(new Animated.Value(0)).current;

  useEffect(() => {
    playOpacityAnimation(opacityAnimation);
    playSizeAnimation(heightAnimation, height);
    playShakeAnimation(shakeAnimation);
  }, [height, heightAnimation, opacityAnimation, shakeAnimation, winner]);

  if (winner === '') {
    return null;
  }

  return (
    <Animated.View
      style={getLineStyle({
        winner,
        length,
        height,
        right: height,
        heightAnimation,
        opacityAnimation,
        shakeAnimation,
      })}
    />
  );
};

export default DiagonalTopRightLine;
