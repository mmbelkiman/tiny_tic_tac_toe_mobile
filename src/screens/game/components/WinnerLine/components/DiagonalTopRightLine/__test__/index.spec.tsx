import {Animated} from 'react-native';
import React from 'react';
import renderer from 'react-test-renderer';
import DiagonalTopRightLine from '../index';

describe('DiagonalTopRightLine Component', () => {
  it('should match the snapshot with a winner', () => {
    const heightAnimation = new Animated.Value(0);
    const opacityAnimation = new Animated.Value(0);
    const shakeAnimation = new Animated.Value(0);

    const tree = renderer
      .create(
        <DiagonalTopRightLine
          length={100}
          height={20}
          color="blue"
          shakeAnimation={shakeAnimation}
          heightAnimation={heightAnimation}
          opacityAnimation={opacityAnimation}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should match the snapshot without a winner', () => {
    const heightAnimation = new Animated.Value(0);
    const opacityAnimation = new Animated.Value(0);
    const shakeAnimation = new Animated.Value(0);

    const tree = renderer
      .create(
        <DiagonalTopRightLine
          length={100}
          height={20}
          color="red"
          opacityAnimation={opacityAnimation}
          shakeAnimation={shakeAnimation}
          heightAnimation={heightAnimation}
        />,
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
