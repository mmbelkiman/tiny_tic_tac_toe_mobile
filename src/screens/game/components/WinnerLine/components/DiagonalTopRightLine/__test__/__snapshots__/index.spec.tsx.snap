// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DiagonalTopRightLine Component should match the snapshot with a winner 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "blue",
      "borderRadius": 10,
      "height": 0,
      "opacity": 0,
      "position": "absolute",
      "right": 20,
      "transform": [
        {
          "translateX": 50,
        },
        {
          "translateY": -10,
        },
        {
          "rotate": "-45deg",
        },
        {
          "translateX": -50,
        },
        {
          "translateY": 10,
        },
        {
          "translateY": 0,
        },
      ],
      "width": 100,
    }
  }
/>
`;

exports[`DiagonalTopRightLine Component should match the snapshot without a winner 1`] = `
<View
  collapsable={false}
  style={
    {
      "backgroundColor": "red",
      "borderRadius": 10,
      "height": 0,
      "opacity": 0,
      "position": "absolute",
      "right": 20,
      "transform": [
        {
          "translateX": 50,
        },
        {
          "translateY": -10,
        },
        {
          "rotate": "-45deg",
        },
        {
          "translateX": -50,
        },
        {
          "translateY": 10,
        },
        {
          "translateY": 0,
        },
      ],
      "width": 100,
    }
  }
/>
`;
